core/src/driver/scheduler/evm.rs

```diff
@@ -32,110 +38,119 @@ impl EvmScheduler {
             driver,
             exchange,
             config,
+            sleep: Box::new(AsyncSleep),
             last_batch: None,
         }
     }
 
     /// Creates a new scheduler with the default configuration.
     #[cfg(test)]
-    pub fn with_defaults(
+    pub fn with_defaults_and_sleep(
         exchange: Arc<dyn StableXContract>,
         driver: Arc<dyn StableXDriver>,
+        sleep: Box<dyn AsyncSleeping>,
     ) -> Self {
-        EvmScheduler::new(exchange, driver, AuctionTimingConfiguration::default())
+        EvmScheduler {
+            driver,
+            exchange,
+            config: AuctionTimingConfiguration::default(),
+            sleep,
+            last_batch: None,
+        }
     }
 
-    /// Runs the scheduler for a single iteration.
-    fn step(&mut self) -> Result<()> {
-        let batch_id = self.current_solving_batch()?;
-        if self
-            .last_batch
-            .map(|last_batch| batch_id <= last_batch)
-            .unwrap_or_default()
-        {
-            debug!("skipping already processed batch {}", batch_id);
-            return Ok(());
-        }
+    /// Gets the current solving batch ID.
+    ///
+    /// This is the current batch ID minus 1, as the current batch ID is the ID
+    /// of the batch that is currently accepting orders.
+    async fn current_solving_batch(&self) -> Result<u32> {
+        Ok(self.exchange.get_current_auction_index().await? - 1)
+    }
 
-        let time_remaining = self.exchange.get_current_auction_remaining_time().wait()?;
-        // NOTE: We need to take into account the asynchronous nature of web3
-        //   and handle the case where we query the batch information right on
-        //   a batch border and the following happens:
-        //     - query batch ID and get `N`
-        //     - block gets mined, batch ID becomes `N+1`
-        //     - query time remaining and get around `300s`
-        //   In order to work around this, we just re-query the batch ID after
-        //   getting the time in the batch to make sure we are using the correct
-        //   batch. If they don't match, we just return to restart the run-loop.
-        let verify_batch_id = self.current_solving_batch()?;
-        if batch_id != verify_batch_id {
-            info!(
-                "batch ID changed during run loop ({} -> {}); retrying",
-                batch_id, verify_batch_id,
-            );
-            return Ok(());
+    async fn wait_for_batch_to_change(&self, batch: u32) -> Result<u32> {
+        loop {
+            let current_batch = self.current_solving_batch().await?;
+            if current_batch != batch {
+                return Ok(current_batch);
+            }
+            self.sleep.sleep(POLL_TIMEOUT).await;
```

Comment:
Should this be a separate component that we can mock inside unit tests? Or can we drive it by mocking the contract directly?

Reply:
It's already being tested so it doesn't need to be a mockable trait. `scheduler_only_runs_next_batch_after_previous_has_finished`